### PR TITLE
Implement ReadAsync on CurlHandler response stream

### LINKED ISSUE #3067: Implement ReadAsync on CurlHandler's response stream

### PR DESCRIPTION
**Pull request by stephentoub**
Today, CurlHandler provides a response content stream that doesn't override ReadAsync.  As such, any consumers reading data using ReadAsync will actually end up blocking a thread pool thread doing a synchronous Read.

This commit fixes that by implementing ReadAsync to be truly asynchronous with no additional threads blocked.  If data is already available, a task is returned immediately.  Otherwise, the reader asynchronously waits for the writer to publish data and then consumes that data.

The overall design of the response stream remains the same as it was.  However, there are some notable changes:
- A SemaphoreSlim is now used instead of a ManualResetEventSlim for the _writerPublishedDataEvent.  This was done because SemaphoreSlim supports both synchronous and asynchronous waiting.
- The _readerRequestingDataEvent was removed, as it was no longer necessary.
- We no longer allocate an UnmanagedMemoryStream per write operation, instead simply storing the IntPtr buffer and long length on the response stream.
- Additional reorganization was done, primarily to share logic between Read and ReadAsync.
- A bunch of comments and debug asserts were added.

A second commit enables the asynchronous wait to be canceled by either the internal shutdown CancellationToken or by the CancellationToken provided by the caller to ReadAsync.

Fixes #3067 
cc: @vijaykota, @kapilash, @davidsh, @pgavlin, @CIPop

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs [hunk 1]
```diff
@@ -53,31 +91,6 @@ internal void SignalFailure(ExceptionDispatchInfo error)
...
 
-        internal unsafe void WaitAndSignalReaders(IntPtr pointer, long length)
-        {
-            try
-            {
-                CheckDisposed();
-                _readerRequestingDataEvent.Wait(Timeout.InfiniteTimeSpan, _cancellationTokenSource.Token);
-                lock (LockObject)
-                {
-                    _innerStream = new UnmanagedMemoryStream((byte*)pointer, length);
```
Comment (vijaykota) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs]: Nit: You can remove reference to System.IO.UnmanagedMemoryStream

Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs]: Thanks, done.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs [hunk 2]
```diff
@@ -126,164 +139,270 @@ public override long Position
...
+        private int ConsumePublishedData(byte[] buffer, int offset, int count)
         {
-            Debug.Assert(Monitor.IsEntered(LockObject), "lock wasn't held");
-            _writerHasDataEvent.Reset();
-            _readerRequestingDataEvent.Set();
+            Debug.Assert(Monitor.IsEntered(LockObject), "Published data must only be accessed while holding the lock");
+
+            // Copy the data into the read buffer
+            int bytesToRead = (int)Math.Min(_availableDataLength, count);
+            Marshal.Copy(_availableData, buffer, offset, bytesToRead);
```
Comment (kapilash) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs]: Do we need to check the cancellation status here again? Is ```Marshal.Copy``` safe  if there was an error  and ```SignalComplete``` was called (via easyRequest.CleanUp()) ?

Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs]: It's safe: we're inside of the lock guarding access to the available data. I don't think checking for cancellation again will achieve much... most of the call sites checked for cancellation just before calling the function, anyway.

Comment (kapilash) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs]: I was worried about the scenario where libcurl may have consumed the native pointer, that it gave us in the curl-callback. I now realize that wont happen. I was wrong.

**Comment by kapilash** [comment]
thanks @stephentoub . LGTM

**Comment by stephentoub** [comment]
Thanks for reviewing, @kapilash!

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs]: I'll add another debug assert to the beginning of the function to reconfirm that there's data available.

**Comment by vijaykota** [comment]
LGTM

**Pull request merged**